repos/system_upgrade/el7toel8/actors/pcidevicesscanner/libraries/pcidevicesscanner.py

```diff
@@ -0,0 +1,71 @@
+import functools
+import os
+import shlex
+import subprocess
+import six
+
+from leapp.libraries.stdlib import call
+from leapp.models import PCIDevices, PCIDevice
+
+
+def aslist(f):
+    ''' Decorator used to convert generator to list '''
+    @functools.wraps(f)
+    def inner(*args, **kwargs):
+        return list(f(*args, **kwargs))
+    return inner
+
+
+def get_from_list(l, idx, default=''):
+    ''' Get item at index from list or return a default '''
+    try:
+        ret = l[idx]
+    except IndexError:
+        ret = default
+
+    return ret
+
+
+@aslist
```

Comment:
This is more like a question out of curiosity. Why is it better to have a function as a generator and convert the output to list using a decorator, instead of return a list by the function right away?

Reply:
For sometime I imagined that PCIDeviceScanner could be just another function inside SystemFacts actor's library. So I kept the code very close to SystemFacts codebase. Not sure the advantages on this case codewise.

Reply:
I like the succinct form of the generators, it reduces the necessary boilerplate and makes the body of the function more readable, i.e.:
```python
def variant_a():
   to_return = []
   for something in alot_of_things:
       to_return.append(operation(something))
   return to_return
```

Versus

```python
@aslist
def variant_b():
   for something in alot_of_things:
       yield operation(something)
```

Reply:
Oh, and why to convert it to a list straight away? Generators, as useful as they might be, they can sometimes bite your back in unexpected ways, consider:

```python
foo = some_generator()
data = {
  'important': [f for f in foo if f.severity > 100],
  'all': list(foo)
}
```

While the `important` field will have it's proper value, the `all` field will be empty since the generator had already been exhausted by that point.